Adds checklists for doing initial PR review. (#8209)
numba/numba: docs/source/developer/contributing.rst

The core developers thank everyone for their understanding with the above!

.. _reviewing_pull_requests:

Reviewing Pull Requests
-----------------------

Help with reviewing pull requests is welcomed. There are time-effort categorised

> review tasks available on this project board: **TODO: ADD LINK**.

Reviewer: ```suggestion
review tasks available on the `Contributor self-service board <https://github.com/orgs/numba/projects/7>`_.
```

Below are some check lists to help with reviewing pull requests that are adding,
or altering, an implementation of some functionality e.g. adding a new NumPy
function or updating some Python built-in. Consider selecting the check list
appropriate to your use case and copying it into the issue as a comment, you can
then check off the boxes as you go. Note that these check lists are a guide,
are not exhaustive, and some parts may not always be applicable!

NumPy addition/change review check list::

    I have checked that:

    * [ ] There is no other Pull Request proposing the same change. (If there is
          please highlight this and one of the core developers will pick it up.)
    * [ ] The proposed additional NumPy functionality is not deprecated by
          NumPy.
    * [ ] Public CI is passing.
    * [ ] The patch does not touch files in `numba.core`.
    * [ ] The patch is implemented using the high-level extension API only
          (https://numba.readthedocs.io/en/stable/extending/high-level.html).
    * [ ] The patch updates the documentation to reflect the new functionality.
          (https://github.com/numba/numba/blob/main/docs/source/reference/numpysupported.rst)

>     * [ ] The patch is line wrapped to 80 characters.

Author: Need to add a note about running a "spell checker", and making sure that NumPy/SciPy/Python are correctly capitalised.

Reviewer 2: hm... can we automate these type of checks as part of CI?

    * [ ] Changes to Python source in the patch are `flake8` compliant.
    * [ ] The implementation of the NumPy function reflects that present in the
        NumPy implementation itself (look in the NumPy repository:
          https://github.com/numpy/numpy)...
        * [ ] The implementation has references to the region of the NumPy code
              base that provides the same functionality. The references are URLs
              of the form:
              `https://github.com/numpy/numpy/tree/<SHA>/path/to/file.py:#<lines>`
        * [ ] The code "looks" reasonably similar to the NumPy implementation.
        * [ ] The code uses the same algorithm as the NumPy implementation, if
              it does not it explains why and this justification is reasonable.
        * [ ] The type inference part of the Numba implementation restricts the
              types to those declared in the NumPy docs (or if not present, what

>               is accepted in the NumPy implementation itself).

Reviewer: Probably worth adding that those restrictions are tested by supplying bogus types and checking that the error message is indeed the expected one.

              different NumPy versions in a clear fashion and uses
              `numba.np.numpy_support.numpy_version` to separate the difference
              in implementations.
    * [ ] The unit tests for the implementation:
        * [ ] Use the same test inputs as the NumPy unit tests for the same
              functionality. If this is not the case there are comments which
              explain why and this justification is reasonable.
        * [ ] The NumPy unit test inputs have reference URLs associated with
              them to track their origin. The references are URLs of the form:
              `https://github.com/numpy/numpy/tree/<SHA>/path/to/file.py:#<lines>`
        * [ ] Use `unittest` style assertions (not pytest style!).
        * [ ] Check that unsupported/invalid input types are handled as expected
              using the `with self.assertRaises()` pattern.
        * [ ] Explicitly check custom error messages are raised appropriately.
        * [ ] Hit all the branches of the proprosed change (both at type
              inference and run time).
        * [ ] Check that numerically awkward values work as expected e.g.

>               `NaN`, `+/-Inf`, `0`, `INT_MAX`, `INT_MIN` etc.

Reviewer 2: I wonder if we should reference Boundary Value Testing in addition to this as this is a subset of BVT. BVT will also include empty containers.
